### PR TITLE
Fix multimodal draft embedding alignment

### LINKED ISSUE #35440: [Bug] `mm_input_embeds` is corrupted before the draft model reads it
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [ ] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

## Summary

`forward_batch.mm_input_embeds` carries the target model's *fused* multimodal input embeddings to
the draft model during speculative decoding (the draft cannot recompute them: at image/video
positions `embed_tokens(input_ids)` returns the placeholder embedding, not the vision features).
Two things make what the draft actually reads wrong:

1. **The field stores a reference to a tensor the target overwrites in place**, so the draft reads
   the target's final hidden states instead of embeddings.
2. **The draft's `input_ids` are rotated left by one; `mm_input_embeds` is not**, so every prefix
   slot is off by one.

Consumers affected: `models/qwen3_5_mtp.py:192`, `models/llama_eagle3.py:209`, i.e. any reader of
the field. Text-only serving, and multimodal without speculative decoding, are unaffected.

## Issue 1 — the tensor is overwritten in place by the target's residual stream

```
managers/mm_utils.py:719      forward_batch.mm_input_embeds = input_embeds   # reference, not a copy
managers/mm_utils.py:730-735  language_model(..., input_embeds=input_embeds) # same tensor
models/qwen3_5.py:1426          hidden_states = input_embeds
layers/communicator.py:620        residual = hidden_states                   # `residual is None` branch
layers/layernorm.py:582             fused_add_rmsnorm(x, residual, w, eps)   # IN-PLACE on residual
```

`fused_add_rmsnorm` writes `residual <- x + residual` back into the caller's storage, and in
layer 0 `residual` *is* `input_embeds`. The embedding buffer is therefore overwritten before
layer 0 even finishes, and the remaining layers keep accumulating into it. Every consumer of the
field runs **after** the target forward, so the draft reads the target's final hidden states.

The pre-allocated-buffer copy at `mm_utils.py:722-724` does not prevent this: `batch.input_embeds`
is only non-`None` when the request itself supplied raw embeddings
(`managers/schedule_batch.py:2574`), which is not the case for an ordinary image/video request.

Measured on one image request (237 placeholder tokens), per-row L2 norm `min/med/max`:

| tensor                  | min  | med   | max        |
|-------------------------|------|-------|------------|
| target side (expected)  | 0.28 | 2.38  | 9.95       |
| draft side (actual)     | 6.96 | 13.26 | **140.47** |

A ±2 offset scan finds no alignment between the two, which rules out "merely misaligned
embeddings" — the draft is reading a different quantity.

## Issue 2 — the left rotation applied to `input_ids` is not applied to `mm_input_embeds`

MTP slot `j` consumes `Emb(t_{j+1})`, and `_draft_extend_for_prefill` rotates `input_ids` left by
one per request accordingly:

```python
# speculative/eagle_worker_v2.py:750-758
tail_tokens = _eagle_prefill_tail_tokens(batch, next_token_ids)
for i, extend_len in enumerate(batch.extend_lens):
    input_ids = batch.input_ids[pt : pt + extend_len]
    new_input_ids[pt : pt + extend_len].copy_(
        torch.cat((input_ids[1:], tail_tokens[i].reshape(1)))
    )
```

`mm_input_embeds` is attached unmodified (`speculative/eagle_worker_v2.py:786-787`) and stays
aligned to the target sequence, while the consumers only patch each request's last row:

```python
# models/qwen3_5_mtp.py:196-204 (models/llama_eagle3.py:211-218 is the same shape)
last_indices = (forward_batch.extend_start_loc + forward_batch.extend_seq_lens - 1).long()
input_embeds[last_indices] = self.model.embed_tokens(input_ids[last_indices])
```

So slot `j` receives `Emb(t_j)` where `Emb(t_{j+1})` is required; only the last slot per request is
correct. Note also that this assignment writes in place into the target's tensor.

### Reproduction

With any VLM plus EAGLE3 or MTP/NEXTN, log per-row L2 norms of the multimodal placeholder rows
(a) in `general_mm_embed_routine` before `language_model(...)`, and (b) in the draft model right
after `input_embeds = forward_batch.mm_input_embeds`. The two disagree by an order of magnitude,
and an offset scan over `d in [-2, +2]` finds no alignment.

### Environment

No special

### PR DESCRIPTION
**Pull request by JustinTong0323**
## Motivation

Fixes #35440.

`forward_batch.mm_input_embeds` is consumed by EAGLE/MTP draft models, but eager target execution can mutate the same fused multimodal embedding tensor in place before the draft reads it. The draft then receives corrupted embeddings; the failure is silent and primarily shows up as lower speculative acceptance.

The separate row-rotation half of the original issue is already on current `main` via #36507 / #37548, so this PR is narrowed to the remaining ownership fix.

## Modifications

- Snapshot fused multimodal embeddings for EAGLE-family speculative decoding only when `forward_batch.input_embeds` is absent (eager, FULL, or TC_PIECEWISE fallback paths).
- Avoid the extra full-size clone on BREAKABLE prefill CUDA graph, where the target mutates the preallocated `forward_batch.input_embeds` buffer and the original fused embedding tensor remains safe.
- Add registered CPU tests for both ownership cases: eager in-place target mutation preserves the draft snapshot, and a static input buffer reuses the source tensor without another clone.

## Validation

- `pre-commit` on changed files: passed.
- Registered CPU test: 2 passed.
- Qwen3.5-27B BF16 + NEXTN/EAGLE on one H200, eager prefill: 512/512 image requests succeeded at concurrency 16; average speculative acceptance length was 2.52.
- The same model with default BREAKABLE prefill CUDA graph: 64/64 image requests succeeded at concurrency 16; average speculative acceptance length was 2.57.
- Full GSM8K (1319 examples, temperature 0.7, top-p 0.8, max 4096, `enable_thinking=false`): 96.82% accuracy, 99.70% stop-rate, 0.30% truncated, 0 errors.

## Related work

#27647 and #27846 fixed per-request final-row replacement for batched multimodal draft prefill. #36507 / #37548 brought the remaining per-request row rotation into `main`. This PR now only fixes the eager snapshot ownership problem and avoids paying the snapshot copy on the BREAKABLE static-buffer path.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #34621669460](https://github.com/sgl-project/sglang/actions/runs/34621669460)
Latest PR Test (Extra): :x: [Run #34621668740](https://github.com/sgl-project/sglang/actions/runs/34621668740)
Latest PR Test (AMD ROCm 10): :x: [Run #34621669484](https://github.com/sgl-project/sglang/actions/runs/34621669484)

### CODE REVIEW

**Comment by bowenwan6** [comment]
Followed the tensor from where it's created to where the draft reads it, and checked the shift indices by hand. The code is clean and the issue is critical. `mm_input_embeds` is the only channel carrying fused multimodal embeddings to a draft head that cannot recompute them, and two independent things were breaking it — the aliasing, and the rotation the shift now adds. They masked each other, which is how the last-token replacement in #27647 / #27846 could land without the rows in front of it ever being rotated. Either way the failure is silent: nothing errors, acceptance length just drops, so it reads as "multimodal is slow" indefinitely.

**One thing I noticed: on BCG the target mutates the static `input_embeds` slot, so the clone looks skippable there.**

`PrefillInputBuffers.create` allocates that slot whenever `is_multimodal`, and on the `BREAKABLE` backend `static_forward_batch` — the batch carrying it — *is* the serving batch handed to `model.forward`. So `forward_batch.input_embeds is not None` a few lines below the clone: `copy_` stages into the slot and the target mutates that, not the tensor `embed_mm_inputs` just returned. Nothing touches the original there, so the clone becomes a second full-size `[num_tokens, hidden]` copy at peak prefill, on top of the `empty_like` in the shift — ~256 MB at 32k tokens × 4096 bf16.

```python
# The target mutates whatever it is handed; the pre-allocated buffer a few lines
# below takes that mutation when it exists, leaving the original untouched.
needs_private_copy = (
    forward_batch.spec_algorithm is not None
    and forward_batch.spec_algorithm.is_eagle()
    and forward_batch.input_embeds is None
)
forward_batch.mm_input_embeds = (
    input_embeds.clone() if needs_private_copy else input_embeds
)
```

`default_prefill_backend()` returns `BREAKABLE` on CUDA and neither the Qwen3-VL nor the Qwen3.5 override changes it, so this is the default path for the models this PR targets. `FULL` runs the tail against the live `forward_batch` and the EAGLE target on `TC_PIECEWISE` falls back to the eager runner, so `input_embeds` is `None` on both and the clone stays load-bearing there.

Other than that, the fix looks correct and well-scoped to me.

**Comment by JustinTong0323** [comment]
/rerun-test test/registered/e2e/models/test_dsa_glm52_pd_mtp_cp_layersplit.py
